### DISCUSSION TITLE
Queue qty

### DISCUSSION THREAD

**Discussion by mauricemullens** [eModbus/eModbus#367] [eModbus applications]
Hi,
is there a way to get the current number of messages that are placed in the queue?
i have a number of Modbus devices connected, which sometimes do not seem to respond, increasing the queue. I want to be able to act on that by either reconnecting of clearing the queue.
Thanks,
Maurice

**Comment by Miq1** [comment]
There is a ``pendingRequests()`` method with the client objects that may be what you are looking for.

**Reply by mauricemullens** [reply]
'class ModbusClient' has no member named 'pendingRequests'. 
i am using ModbusClientTCPasync. How can i reach pendingRequests which is in ModbusClientRTU?

**Reply by Miq1** [reply]
It is in ``ModbusClientTCP`` as well. Looks like @bertmelis decided to not implement it (yet) in the async code. Sorry...

There is a ``std::queue`` of ``RequestEntry`` objects inside there. You would need to add some function to export the ``std::queue.size()`` method out of it.

**Reply by mauricemullens** [reply]
i added the code to ModbusClientTCPasync. Working well

.h
uint32_t getQueueCount();
uint32_t queue_cnt = 0;

.cpp
uint32_t ModbusClientTCPasync::getQueueCount()
{
    return queue_cnt;
}

// addToQueue: send freshly created request to queue
bool ModbusClientTCPasync::addToQueue(int32_t token, ModbusMessage request, bool syncReq)
{

    queue_cnt = txQueue.size() + rxQueue.size();
    // Did we get one?
    if (request) ... etc

**Reply by mauricemullens** [reply]
BTW there is another small error:
In file included from src/main.cpp:12:
.pio/libdeps/wt32-eth01/eModbus/src/ModbusClientTCP.h:22: warning: "DEFAULTTIMEOUT" redefined
 #define DEFAULTTIMEOUT 2000
 
In file included from src/devices.h:5,
                 from src/main.cpp:3:
.pio/libdeps/wt32-eth01/eModbus/src/ModbusClientTcpAsync.h:25: note: this is the location of the previous definition
 #define DEFAULTTIMEOUT 10000
 
i made them both the same now, error gone. You might want to change this in the code?

**Reply by Miq1** [reply]
Uh, there is something odd going on. You may not have both the synchronous and async code in one application - it looks like both are mixed in your code.

**Reply by Miq1** [reply]
> ```
> queue_cnt = txQueue.size() + rxQueue.size();
> // Did we get one?
> if (request) ... etc
> ```

Makes no sense to add requests and responses here, I would suggest using the TX queue only.

**Reply by mauricemullens** [reply]
> Uh, there is something odd going on. You may not have both the synchronous and async code in one application - it looks like both are mixed in your code.

yes i was testing with modbustcp.h. I changed it back and no problem anymore

**Comment by mauricemullens** [comment]
this is a bit stupid maybe, but i added the code in the .pio dir, but that was (obviously) overwritten. But i cant find the original files linked in platformio. Do i need to fork it and add it manually? Or do a change request? Or do you know perhaps where the original fiels are stored? Searched the whole drive (OSX) but cant find it anywhere :-(

**Reply by Miq1** [reply]
``.pio/lib_deps/eModbus/src`` AFAIRC
